src/main/kotlin/org/rust/lang/core/psi/RustDocAndAttributeOwner.kt

```diff
@@ -35,16 +36,55 @@ class QueryAttributes(private val attributes: List<RustAttrElement>) {
             .mapNotNull { it.litExpr?.stringLiteralValue }
             .singleOrNull()
 
-    val metaItems: List<RustMetaItemElement>
+    val metaItems: Sequence<RustMetaItemElement>
         get() = attributes.mapNotNull { it.metaItem }
+}
+
+/**
+ * Iterator that walks through both inner and outer attributes without allocating extra collections.
+ * Inner attributes have priority, so they are iterated first.
+ */
+class RustAttributeIterator (
+    val owner: RustDocAndAttributeOwner
+) : Iterator<RustAttrElement> {
```

Comment:
I think you can more shortly implement a lazy version with sequence API, without fallback to the `Iterator`:

```Kotlin
val RustDocAndAttributeOwner.allAttributes: Sequence<RustAttrElement>
    get() = Sequence { (this as? RustOuterAttributeOwner)?.outerAttrList.orEmpty().iterator() } +
        Sequence { (this as? RustInnerAttributeOwner)?.innerAttrList.orEmpty().iterator() }
```

:+1: for switching to sequences from lists! Unfortunately base Grammar Kit APIs allocate collections, but this is something we can work around if necessary.